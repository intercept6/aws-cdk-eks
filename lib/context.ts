import { Environment, StackProps } from "@aws-cdk/core";

export interface Context {
  env: Environment;
  domain: string;
}
