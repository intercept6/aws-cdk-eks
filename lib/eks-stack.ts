import { Construct, Stack, StackProps, CfnOutput } from "@aws-cdk/core";
import { Cluster } from "@aws-cdk/aws-eks";
import {
  InstanceClass,
  InstanceSize,
  InstanceType,
  IVpc
} from "@aws-cdk/aws-ec2";
import { ManagedPolicy } from "@aws-cdk/aws-iam";

interface EksStackProps extends StackProps {
  vpc: IVpc;
}

export class EksStack extends Stack {
  public readonly cluster: Cluster;

  constructor(scope: Construct, id: string, props: EksStackProps) {
    super(scope, id, props);

    this.cluster = new Cluster(this, "Eks", {
      vpc: props.vpc,
      version: "1.13"
    });
    const asg = this.cluster.addCapacity("WorkerNodes", {
      instanceType: InstanceType.of(InstanceClass.M5, InstanceSize.LARGE),
      minCapacity: 2,
      maxCapacity: 4,
      desiredCapacity: 2
    });
    asg.addUserData(
      "cd /tmp\n" +
        "yum install -y https://s3.amazonaws.com/ec2-downloads-windows/SSMAgent/latest/linux_amd64/amazon-ssm-agent.rpm\n" +
        "systemctl start amazon-ssm-agent"
    );

    const managedPolicyNames = [
      "service-role/AmazonEC2RoleforSSM",
      "ElasticLoadBalancingFullAccess",
      "AutoScalingFullAccess"
    ];
    for (let managedPolicyName of managedPolicyNames) {
      asg.role.addManagedPolicy(
        ManagedPolicy.fromAwsManagedPolicyName(managedPolicyName)
      );
    }

    // Output
    new CfnOutput(this, "EksClusterName", {
      value: this.cluster.clusterName
    });
    new CfnOutput(this, "EksWorkerNodesInstanceRoleARN", {
      value: asg.role.roleArn
    });
  }
}
