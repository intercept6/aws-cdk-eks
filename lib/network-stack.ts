import { Stack, StackProps, Construct } from "@aws-cdk/core";
import { Vpc, SubnetType } from "@aws-cdk/aws-ec2";
import { Certificate } from "@aws-cdk/aws-certificatemanager";

export class NetworkStack extends Stack {
  public readonly vpc: Vpc;

  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    // TODO: Publicサブネットにタグを付ける(kubernetes.io/role/elb:1)
    this.vpc = new Vpc(this, "VPC", {
      subnetConfiguration: [
        { name: "Public", subnetType: SubnetType.PUBLIC, cidrMask: 20 },
        { name: "Private", subnetType: SubnetType.PRIVATE, cidrMask: 20 },
        { name: "Isolated", subnetType: SubnetType.ISOLATED, cidrMask: 20 }
      ]
    });

    // TODO: 証明書をCDKで作成する
    // const cert = new Certificate(this, "Certificate", {
    //   domainName: 'rkato.classmethod.info',
    //   subjectAlternativeNames: '*.rkato.classmethod.info',
    // });
  }
}
