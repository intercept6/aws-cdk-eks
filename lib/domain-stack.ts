import {CfnOutput, Construct, Stack, StackProps, Tag} from "@aws-cdk/core";
import {DnsValidatedCertificate, ValidationMethod} from "@aws-cdk/aws-certificatemanager";
import {HostedZone} from "@aws-cdk/aws-route53";
import {Context} from "./context";


export class DomainStack extends Stack {

    constructor(scope: Construct, id: string, props?: StackProps) {
        super(scope, id, props);
        const stage: string = this.node.tryGetContext("stage");
        const params: Context = this.node.tryGetContext(stage);


        // Generate TLS certificate that is DNS verified by Route53
        const hostedZone = HostedZone.fromLookup(this, "HostedZone", {
            domainName: params.domain
        });
        const acm = new DnsValidatedCertificate(this, "Certificate", {
            domainName: `*.${params.domain}`,
            subjectAlternativeNames: [params.domain],
            validationMethod: ValidationMethod.DNS,
            region: params.env.region,
            hostedZone
        });
        // acm.node.applyAspect(new Tag("Name", "aws-cdk-eks"));

        // Output
        new CfnOutput(this, "AcmArn", {
            value: acm.certificateArn
        });
    }
}
