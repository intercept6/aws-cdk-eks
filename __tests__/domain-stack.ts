import { App } from "@aws-cdk/core";
import { SynthUtils } from "@aws-cdk/assert";
import { DomainStack } from "../lib/domain-stack";
import { Context } from "../lib/context";

describe("domain", () => {
  test("default", () => {
    const app = new App();
    const stage = "test";
    app.node.setContext("stage", stage);
    const params: Context = {
      env: {
        account: "111111111111",
        region: "ap-northeast-1"
      },
      domain: "example.com"
    };
    app.node.setContext(stage, params);

    const stack = new DomainStack(app, "TestDomainStack", {
      env: params.env
    });
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });
});
