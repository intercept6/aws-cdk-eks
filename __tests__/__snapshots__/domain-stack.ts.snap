// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`domain default 1`] = `
Object {
  "Outputs": Object {
    "AcmArn": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "CertificateCertificateRequestorResource2890C6B7",
          "Arn",
        ],
      },
    },
  },
  "Parameters": Object {
    "CertificateCertificateRequestorFunctionCodeArtifactHashB6FA7A49": Object {
      "Description": "Artifact hash for asset \\"TestDomainStack/Certificate/CertificateRequestorFunction/Code\\"",
      "Type": "String",
    },
    "CertificateCertificateRequestorFunctionCodeS3BucketB11F2FC2": Object {
      "Description": "S3 bucket for asset \\"TestDomainStack/Certificate/CertificateRequestorFunction/Code\\"",
      "Type": "String",
    },
    "CertificateCertificateRequestorFunctionCodeS3VersionKeyA887D87E": Object {
      "Description": "S3 key for asset version \\"TestDomainStack/Certificate/CertificateRequestorFunction/Code\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "CertificateCertificateRequestorFunction5E845413": Object {
      "DependsOn": Array [
        "CertificateCertificateRequestorFunctionServiceRoleDefaultPolicy3C8845BC",
        "CertificateCertificateRequestorFunctionServiceRoleC04C13DA",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "CertificateCertificateRequestorFunctionCodeS3BucketB11F2FC2",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "CertificateCertificateRequestorFunctionCodeS3VersionKeyA887D87E",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "CertificateCertificateRequestorFunctionCodeS3VersionKeyA887D87E",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.certificateRequestHandler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "CertificateCertificateRequestorFunctionServiceRoleC04C13DA",
            "Arn",
          ],
        },
        "Runtime": "nodejs8.10",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CertificateCertificateRequestorFunctionServiceRoleC04C13DA": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CertificateCertificateRequestorFunctionServiceRoleDefaultPolicy3C8845BC": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "acm:RequestCertificate",
                "acm:DescribeCertificate",
                "acm:DeleteCertificate",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "route53:GetChange",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "route53:changeResourceRecordSets",
              "Effect": "Allow",
              "Resource": "arn:aws:route53:::hostedzone/DUMMY",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CertificateCertificateRequestorFunctionServiceRoleDefaultPolicy3C8845BC",
        "Roles": Array [
          Object {
            "Ref": "CertificateCertificateRequestorFunctionServiceRoleC04C13DA",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CertificateCertificateRequestorResource2890C6B7": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "DomainName": "*.example.com",
        "HostedZoneId": "DUMMY",
        "Region": "ap-northeast-1",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CertificateCertificateRequestorFunction5E845413",
            "Arn",
          ],
        },
        "SubjectAlternativeNames": Array [
          "example.com",
        ],
      },
      "Type": "AWS::CloudFormation::CustomResource",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
